### DISCUSSION TITLE
Run time configuration using cloud-init not working

### DISCUSSION THREAD

**Discussion by ventureElen** [rancher/elemental-toolkit#1838] [Q&A]
I have created qcow2 image of opensuse following the steps: https://rancher.github.io/elemental-toolkit/docs/creating-derivatives/build_disk/
Used following image as vanilla elemental image for opensuse: ghcr.io/rancher/elemental-toolkit/elemental-green:v0.10.7-g3e4a3c56.
I used Virtual Box on windows 11 to create VM using the qcow2 image. i m able login to VM using root/cos. then i tried to install net-tools package using zypper but it gave the following error:

elemental:~ # zypper in -y net-tools
The target filesystem is mounted as read-only. Please make sure the target filesystem is writeable.

Following the documentation, i have created an yaml file and placed it on /usr/local/cloud-config/:

install.yaml contents:

stages:
name: "Install something"
stages:
   after-upgrade-chroot:
     - commands:
        - zypper in -y net-tools
   after-reset-chroot:
     - commands:
        - zypper in -y net-tools
   after-deploy-chroot:
     - commands:
        - zypper in -y net-tools
   after-install-chroot:
     - commands:
        - zypper in -y net-tools

upon reboot, netstat isn't installed. 

I tried placing the same file on /oem directory but still it didn't work. i also tried to overwrite the default username and password by copying accounting file to /usr/local/cloud-config but is it still accepting root/cos. injecting of environment variables also didn't work. in short, placement of new yaml files in either /oem or /usr/local/cloud-config has no impact and the VM seems to be rebooting from default cloud config files only.

Also elemental install/elemental upgrade are not working with following error:

elemental:/oem # elemental install
\INFO[2023-09-26T14:34:49Z] Starting elemental version v0.10.7
INFO[2023-09-26T14:34:49Z] reading configuration form '/etc/elemental'
ERRO[2023-09-26T14:34:49Z] invalid install command setup undefined system source to install
Error: undefined system source to install
elemental:/oem # \

Please let me know what am i doing wrong?

Documentation says that a derivative is a simple container image. does that mean that iso or qcow2 or vdi files created from the container images are not elemental derivates and hence run time configuration won't work? please clarify.

**Comment by frelon** [comment]
Hi @ventureElen! Great to see you are trying out elemental!

The provided elemental-green example is not really production ready, as you have seen. I would encourage you to copy the [example](https://github.com/rancher/elemental-toolkit/blob/main/examples/green/Dockerfile).

In order to install the net-tools package it can be added to the [list of packages](https://github.com/rancher/elemental-toolkit/blob/2717827f8c1f446feeee300038e7b8162c853388/examples/green/Dockerfile#L15) in the Dockerfile

And if you don't want the default root password set you can change the invocation of `elemental init` on [this line] to the following:
`elemental init -f immutable-rootfs,grub-config,elemental-setup,dracut-config,cloud-config-essentials` and then COPY in your own accounting.yaml file into `/system/oem` or `/usr/local/cloud-config` that ensures your user exists.

Let me know if you run into any troubles!

**Reply by ventureElen** [reply]
Ok so in dockerfile after running elemental init, i can copy any number of files into /system/oem to overwrite default functionality right?

**Reply by frelon** [reply]
Yes, that is correct.

**Comment by kkaempf** [comment]
Keep in mind that the operating system is packaged in a container image. As such its immutable and you can't install/update/remove packages inside the running OS. 😉

**Reply by ventureElen** [reply]
I have tried removing vim package by placing an yaml file into /usr/local/cloud-config. and then triggered elemental upgrade. it gave some error so then i specified the container image of the target system (docker image from which iso/qcow2 images r created) in system.uri and once upgrade is completed, i rebooted the VM and vim package was gone. so it seems to be working but elemental upgrade command takes time even if image isn't changed. is there any way to install/remove a package without doing complete upgrade ?

**Reply by ventureElen** [reply]
Like instead of specifying system image, can we give config directory or file path in elemental upgrade command which won't do full fledged upgrade (by downloading new image or using existing image) but just apply changes on top of running OS?

**Comment by kkaempf** [comment]
> is there any way to install/remove a package without doing complete upgrade ?

No.
In the same way it isn't possible to change the contents of a running container.
You need to rebuild the container image and spawn a new instance.

**Comment by kkaempf** [comment]
> elemental upgrade command takes time even if image isn't changed.

Upgrade compares the `TIMESTAMP` setting in `/etc/os-release` to decide if images are different.

**Comment by kkaempf** [comment]
> Like instead of specifying system image, can we give config directory or file path in elemental upgrade command which won't do full fledged upgrade (by downloading new image or using existing image) but just apply changes on top of running OS?

No.
Elemental provides a _declarative_ approach to os management, not an _imperative_ one.
You need to switch viewpoints 😉. Away from a traditional OS, where you can log in and make changes, to a container workload which is defined by a build process outside of the workload itself.
